spring-integration-amqp/src/main/java/org/springframework/integration/amqp/outbound/AsyncOutboundGateway.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.integration.amqp.outbound;
+
+import org.springframework.amqp.core.AmqpMessageReturnedException;
+import org.springframework.amqp.rabbit.core.AsyncRabbitTemplate;
+import org.springframework.amqp.rabbit.core.AsyncRabbitTemplate.RabbitMessageFuture;
+import org.springframework.amqp.rabbit.support.CorrelationData;
+import org.springframework.amqp.support.converter.MessageConverter;
+import org.springframework.amqp.support.converter.SimpleMessageConverter;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHandlingException;
+import org.springframework.messaging.MessagingException;
+import org.springframework.messaging.support.ErrorMessage;
+import org.springframework.util.concurrent.ListenableFutureCallback;
+
+import reactor.core.support.Assert;
+
+/**
+ * An outbound gateway where the sending thread is released immediately and the reply
+ * is sent on the async template's listener container thread.
+ *
+ * @author Gary Russell
+ * @since 4.3
+ *
+ */
+public class AsyncOutboundGateway extends AbstractAmqpOutboundEndpoint {
```

Comment:
`Async**Amqp**OutboundGateway`?
Without the protocol it sounds very common, but it isn't - just for AMQP.